capsules/src/net/network_capabilities.rs

```diff
@@ -0,0 +1,109 @@
+use crate::net::ipv6::ip_utils::IPAddr;
```

Comment:
Hi Brad, thanks for the feedback -- I'll definitely add more documentation. I can explain a bit more regarding the visibility capabilities (e.g. UdpVisCap, IpVisCap). Phil mentioned that layer separation is advantageous in that we can guarantee that only UDP-related code touches the UDP-relevant aspects of the network capability (accordingly for IP-related code) while we can pass the same network capability up and down the networking stack. This is more to enforce layering separation within the code as opposed to prevent against attacks that can break layering. Does this make sense?

Reply:
The idea of enforcing layer separation makes sense. I'm not necessarily sure it is helpful for writing code, but I'm also not sure it isn't and therefore definitely seems worth experimenting with.

Is there a way to include those capabilities in the networking capsule folder? I think having one capability in the kernel makes sense, but if all others can be included in the net folder then that provides better encapsulation for an experimental interface and will allow for faster development in the future.

Reply:
I can move those over to the networking capsule folder -- also, another rationale for having the layering capabilities would be to enforce principle of least privilege.

Reply:
Could I ask what the one capability in the kernel would be? (I assume you'd be meaning the creation capability?)

Reply:
Also, the issue with moving the trait-based capabilities to the networking capsule folder is that UdpVisCap and IpVisCap need to use the unsafe keyword in their declaration as traits (and this isn't allowed in the networking capsule folder).

Reply:
I was thinking there would be one networking configuration capability in the kernel, and then that would be required to create all of the other networking related capabilities used throughout the stack. The advantage of this approach would be that the network stack could iterate much more quickly (it is easier to merge PRs if they do not modify the kernel crate) with new capabilities.

The downside is the kernel-level capability cannot be stored in the networking stack, because then any code could make a networking-level capability. I don't think this would currently be an issue based on a quick look at the code, but might have effects I haven't considered.

Reply:
I see -- so then the NetworkCapabilityCreationCapability would also be used to instantiate the visibility capabilities? (and these visibility capabilities would no longer be implemented as unsafe traits?)

Reply:
Correct, like the NetworkCapability capability.

Reply:
> One worry: any holder of the creation capability would be able to make arbitrary capabilities for themselves -- we might want it so that regular capsules should never hold onto the creation capability and would only hold onto NetworkCapability and the visibility capabilities -- thoughts on this and how we might explicitly enforce it?

I think this is the key question. My hunch is that since capabilities are meant to be enforced at compile time through function signatures, then there won't be a need to store a copy of the creation capability. But we might just have to experiment with it.

Reply:
(sorry I've been MIA, I was traveling last week) What would you say to the following approach -- we have only one added capability to the kernel -- a network creation capability -- this capability may be used to create the visibility capabilities *as well as* the NetworkCapability objects. This way, we no longer need the static capabilities.

Reply:
Update: I've reworked the code to move the Udp/Ip visibility capabilities out of the kernel (and into the network folder). I've also removed the static capabilities.